packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -167,20 +60,325 @@ export const getOutputAmounts = async (params: GetOutputAmountsParams) => {
+    const amountsWithCredentials = sorted.map(amount => ({
+        amount,
+        credentials: findCredentialsForTarget(
+            amount,
+            amountCredentials,
+            roundParameters.maxAmountCredentialValue,
+        ),
+    }));
+
+    // get first available set and spilt in to output and change
+    const amountPair = amountsWithCredentials.find(c => c.credentials);
+    const vsizePair = findCredentialsForTarget(
+        outputSize,
+        vsizeCredentials,
+        round.roundParameters.maxVsizeCredentialValue,
+    );
+
+    if (amountPair?.credentials && vsizePair) {
+        const availableAmount = sumCredentials(amountPair.credentials);
+        const availableVsize = sumCredentials(vsizePair);
+        const joined = await credentialIssuance({
+            ...params,
+            amountToRequest: [amountPair.amount, availableAmount - amountPair.amount],
+            amountCredentials: amountPair.credentials,
+            vsizeToRequest: [outputSize, availableVsize - outputSize],
+            vsizeCredentials: vsizePair,
+        });
+
+        // create Bob
+        const result = params.result.concat({
+            accountKey: params.accountKey,
+            amount: amountPair.amount,
+            amountCredentials: joined.output.amountCredentials,
+            vsizeCredentials: joined.output.vsizeCredentials,
+        });
+
+        // remove amount from list
+        const amountIndex = amounts.findIndex(a => a === amountPair.amount);
+        const updatedAmounts = amounts.slice(0);
+        updatedAmounts.splice(amountIndex, 1);
+
+        // update credentials list:
+        // - remove joined Credentials used by credentialIssuance process
+        // - add **change** Credentials back to available stack
+        const updatedAmountCredentials = amountCredentials
+            .filter(cre => !amountPair.credentials?.includes(cre))
+            .concat(joined.change.amountCredentials);
+        const updatedVsizeCredentials = vsizeCredentials
+            .filter(cre => !vsizePair.includes(cre))
+            .concat(joined.change.vsizeCredentials);
+
+        if (updatedAmounts.length === 0) {
+            // no more amounts, check whats left
+            const amountDust = sumCredentials(updatedAmountCredentials);
+            const vsizeDust = sumCredentials(updatedVsizeCredentials);
+
+            options.log(`Amount dust: ${amountDust}`);
+            options.log(`Vsize dust: ${vsizeDust}`);
+            options.log('Decomposition completed');
+
+            return result.sort((a, b) => (a.amount > b.amount ? -1 : 1));
+        }
+
+        // try to create another output
+        return createOutputsCredentials({
+            ...params,
+            amounts: updatedAmounts,
+            amountCredentials: updatedAmountCredentials,
+            vsizeCredentials: updatedVsizeCredentials,
+            result,
+        });
+    }
+
+    const amountToJoin = findCredentialsForTarget(
```

Comment:
If I was right above, these `findCredentialsForTarget` calls are unnecessary.

Reply:
this will be called if `amountPair` **or** `vsizePair` was not found.
in that case we want to join **any** credentials for both amount and vsize assuming that `findCredentialsForTarget` will pick the best credentials for target = 0